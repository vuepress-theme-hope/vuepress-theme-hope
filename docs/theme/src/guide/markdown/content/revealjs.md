---
title: Presentation
icon: person-chalkboard
category:
  - Markdown
tag:
  - Markdown
  - Slides
---

Add presentation in your VuePress site via Reveal.js.

<!-- more -->

## Settings

Install `@vuepress/plugin-revealjs` in your project:

::: code-tabs#shell

@tab pnpm

```bash
pnpm add -D @vuepress/plugin-revealjs@next
```

@tab yarn

```bash
yarn add -D @vuepress/plugin-revealjs@next
```

@tab npm

```bash
npm i -D @vuepress/plugin-revealjs@next
```

:::

Then enabling via:

```ts {7} title=".vuepress/config.ts"
import { defineUserConfig } from "vuepress";
import { hopeTheme } from "vuepress-theme-hope";

export default defineUserConfig({
  theme: hopeTheme({
    markdown: {
      revealjs: true,
    },
  }),
});
```

## Slide Syntax

- Use `---` to split slides
- Use `--` to split the slides second time (vertical display)

```md
@slidestart

<!-- slide1 -->

---

<!-- slide2 -->

---

<!-- slide3 -->

@slideend
```

::: details A basic demo

@slidestart

## Slide Title

A paragraph with some text and a [link](https://mister-hope.com)

---

## Highlight

```js [2-4|1-5]
const add = (a, b) => {
  if (typeof b === "undefined") return a + 1;

  return a + b;
};
```

@slideend

````md
@slidestart

## Slide Title

A paragraph with some text and a [link](https://mister-hope.com)

---

## Highlight

```js [2-4|1-5]
const add = (a, b) => {
  if (typeof b === "undefined") return a + 1;

  return a + b;
};
```

@slideend
````

:::

By default, we use `auto` theme to render the presentation, but you can also use other themes with `@slidestart THEME_NAME`.

You can enable the following themes in reveal.js via `themes` in plugin options:

- `auto` (Default)
- `black`
- `white`
- `league`
- `beige`
- `sky`
- `night`
- `serif`
- `simple`
- `solarized`
- `blood`
- `moon`

For the appearance of each theme, see [Themes demo][revealjs-themes-demo].

## Demo

::: md-demo Simple presentation demo

@slidestart

## Slide 1

A paragraph with some text and a [link](https://mister-hope.com)

---

## Slide 2

- Item 1
- Item 2

---

## Slide 3.1

```js
const a = 1;
```

--

## Slide 3.2

$$
J(\theta_0,\theta_1) = \sum_{i=0}
$$

@slideend

:::

::: info

For detailed demo, please see [revealjs-demo][revealjs-demo].

:::

## Customize Reveal.js

### Built-in Plugins

You can enable built-in plugins in reveal.js via `plugins` in plugin options. It accepts an array of the following plugin names:

- `highlight`
- `math`
- `search`
- `notes`
- `zoom`

::: note

`markdown` plugin is enabled anyway to support markdown grammar.

:::

### Advanced Configuration

You can also import and call `defineRevealJsConfig` in [client config file](../../../cookbook/vuepress/config.md#client-config-file) to customize reveal.js:

```js title=".vuepress/client.js"
import { defineRevealJsConfig } from "@vuepress/plugin-revealjs/client";

defineRevealJsConfig({
  // reveal.js options here
});
```

::: note

Reveal.js also provides [more plugins](https://github.com/hakimel/reveal.js/wiki/Plugins,-Tools-and-Hardware), you can add them via `plugin` option in `defineRevealJsConfig`. Built-in plugins you request at node side will be added automatically.

:::

### Per Page Configuration

You can also set `revealJs` to pass options to reveal.js per page in frontmatter.

For reveal.js options, see [reveal.js config](https://revealjs.com/config/). For reveal.js usage, see [reveal.js documentation](https://revealjs.com/)

[revealjs-demo]: https://ecosystem.vuejs.press/plugins/markdown/revealjs/demo.html
[revealjs-themes-demo]: https://ecosystem.vuejs.press/plugins/markdown/revealjs/themes.html
