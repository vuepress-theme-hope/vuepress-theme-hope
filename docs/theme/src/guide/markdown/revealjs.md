---
title: Presentation
icon: person-chalkboard
category:
  - Markdown
tag:
  - Markdown
  - Slides
---

<!-- @include: @md-enhance/guide/content/revealjs/README.md#before -->

::: code-tabs#language

@tab TS

```ts {8-10} title=".vuepress/config.ts"
import { defineUserConfig } from "vuepress";
import { hopeTheme } from "vuepress-theme-hope";

export default defineUserConfig({
  theme: hopeTheme({
    plugins: {
      mdEnhance: {
        revealJs: true,
      },
    },
  }),
});
```

@tab JS

```js {7-9} title=".vuepress/config.js"
import { hopeTheme } from "vuepress-theme-hope";

export default {
  theme: hopeTheme({
    plugins: {
      mdEnhance: {
        revealJs: true,
      },
    },
  }),
};
```

:::

<!-- @include: @md-enhance/guide/content/revealjs/README.md#options -->

For the appearance of each theme, see <ProjectLink name="md-enhance" path="/guide/content/revealjs/themes.html">Themes demo</ProjectLink>.

## Demo

::: md-demo Simple presentation demo

@slidestart

## Slide 1

A paragraph with some text and a [link](https://mister-hope.com)

---

## Slide 2

- Item 1
- Item 2

---

## Slide 3.1

```js
const a = 1;
```

--

## Slide 3.2

$$
J(\theta_0,\theta_1) = \sum_{i=0}
$$

@slideend

:::

::: info

For detailed demo, please see <ProjectLink name="md-enhance" path="/guide/content/revealjs/demo.html">Presentation demo</ProjectLink>

:::

<!-- @include: @md-enhance/guide/content/revealjs/README.md#customize -->
