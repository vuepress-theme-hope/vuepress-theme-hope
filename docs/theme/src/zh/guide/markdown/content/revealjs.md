---
title: 幻灯片
icon: person-chalkboard
category:
  - Markdown
tag:
  - Markdown
  - 幻灯片
---

在你的 VuePress 站点中添加幻灯片。

<!-- more -->

## 配置

在你的项目中安装 `@vuepress/plugin-revealjs`:

::: code-tabs#shell

@tab pnpm

```bash
pnpm add -D @vuepress/plugin-revealjs@next
```

@tab yarn

```bash
yarn add -D @vuepress/plugin-revealjs@next
```

@tab npm

```bash
npm i -D @vuepress/plugin-revealjs@next
```

:::

之后启用它:

```js {6} title=".vuepress/config.js"
import { hopeTheme } from "vuepress-theme-hope";

export default {
  theme: hopeTheme({
    plugins: {
      revealjs: true,
    },
  }),
};
```

## 幻灯片语法

- 使用 `---` 分割幻灯片
- 使用 `--` 对幻灯片进行二次分割(垂直显示)

```md
@slidestart

<!-- slide1 -->

---

<!-- slide2 -->

---

<!-- slide3 -->

@slideend
```

::: details 示例

@slidestart

## 幻灯片标题

一个拥有文字和 [链接](https://mister-hope.com) 的段落

---

## 代码高亮

```js [2-4|1-5]
const add = (a, b) => {
  if (typeof b === "undefined") return a + 1;

  return a + b;
};
```

@slideend

````md
@slidestart

## 幻灯片标题

一个拥有文字和 [链接](https://mister-hope.com) 的段落

---

## 代码高亮

```js [2-4|1-5]
const add = (a, b) => {
  if (typeof b === "undefined") return a + 1;

  return a + b;
};
```

@slideend
````

:::

默认情况下，我们使用 `auto` 主题来渲染幻灯片，你也可以通过 `@slidestart 主题名称` 使用其他主题。

你可以通过插件选项中的 `themes` 启用以下主题:

- `auto` (默认)
- `black`
- `white`
- `league`
- `beige`
- `sky`
- `night`
- `serif`
- `simple`
- `solarized`
- `blood`
- `moon`

各主题的外观，详见 [幻灯片主题][revealjs-themes-demo]。

## 演示

::: md-demo 简单幻灯片演示

@slidestart

## 幻灯片 1

一个有文字和 [链接](https://mister-hope.com) 的段落

---

## 幻灯片 2

- 项目 1
- 项目 2

---

## 幻灯片 3.1

```js
const a = 1;
```

--

## 幻灯片 3.2

$$
J(\theta_0,\theta_1) = \sum_{i=0}
$$

@slideend

:::

::: info

详细与完整的演示详见 [幻灯片演示][revealjs-demo]。

:::

## 自定义 Reveal.js

### 内置插件

你可以通过插件选项中的 `plugins` 启用 reveal.js 中的内置插件。它接受以下插件名称的数组:

- `highlight`
- `math`
- `search`
- `notes`
- `zoom`

::: note

为了支持 Markdown 语法，我们总会启用 `markdown` 插件。

:::

### 高级配置

你也可以在[客户端配置文件](../../../cookbook/vuepress/config.md#客户端配置文件) 中导入并调用 `defineRevealJsConfig` 来自定义 reveal.js:

```ts title=".vuepress/client.ts"
import { defineRevealJsConfig } from "@vuepress/plugin-revealjs/client";

defineRevealJsConfig({
  // 在此设置 reveal.js 选项
});
```

::: note

Reveal.js 还提供了[更多的插件](https://github.com/hakimel/reveal.js/wiki/Plugins,-Tools-and-Hardware)，你可以通过 `plugin` 选项在 `defineRevealJsConfig` 中添加它们，这种自定义行为不会影响你声明的内置插件。

:::

### 页面级配置

你也可以在 Frontmatter 设置 `revealJs` 以设置特定页面的 reveal.js 选项。

Reveal.js 选项，请参见[reveal.js config](https://revealjs.com/config/)，Reveal.js 用法，请参阅 [reveal.js 文档](https://revealjs.com/)。

[revealjs-demo]: https://ecosystem.vuejs.press/zh/plugins/markdown/revealjs/demo.html
[revealjs-themes-demo]: https://ecosystem.vuejs.press/zh/plugins/markdown/revealjs/themes.html
