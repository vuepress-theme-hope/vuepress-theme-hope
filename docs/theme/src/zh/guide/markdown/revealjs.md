---
title: 幻灯片
icon: person-chalkboard
category:
  - Markdown
tag:
  - Markdown
  - 幻灯片
---

<!-- @include: @md-enhance/zh/guide/content/revealjs/README.md#before -->

::: code-tabs#language

@tab TS

```ts {8-10} title=".vuepress/config.ts"
import { defineUserConfig } from "vuepress";
import { hopeTheme } from "vuepress-theme-hope";

export default defineUserConfig({
  theme: hopeTheme({
    plugins: {
      mdEnhance: {
        revealJs: true,
      },
    },
  }),
});
```

@tab JS

```js {7-9} title=".vuepress/config.js"
import { hopeTheme } from "vuepress-theme-hope";

export default {
  theme: hopeTheme({
    plugins: {
      mdEnhance: {
        revealJs: true,
      },
    },
  }),
};
```

:::

<!-- @include: @md-enhance/zh/guide/content/revealjs/README.md#options -->

各主题的外观，详见 <ProjectLink name="md-enhance" path="/zh/guide/content/revealjs/themes.html">幻灯片主题</ProjectLink>。

## 演示

::: md-demo 简单幻灯片演示

@slidestart

## 幻灯片 1

一个有文字和 [链接](https://mister-hope.com) 的段落

---

## 幻灯片 2

- 项目 1
- 项目 2

---

## 幻灯片 3.1

```js
const a = 1;
```

--

## 幻灯片 3.2

$$
J(\theta_0,\theta_1) = \sum_{i=0}
$$

@slideend

:::

::: tip

详细与完整的演示详见 <ProjectLink name="md-enhance" path="/zh/guide/content/revealjs/demo.html">幻灯片演示</ProjectLink>。

:::

<!-- @include: @md-enhance/zh/guide/content/revealjs/README.md#customize -->
