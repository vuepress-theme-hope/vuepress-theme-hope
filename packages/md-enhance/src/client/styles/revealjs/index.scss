@use "@temp/md-enhance/revealjs-theme.scss";
@use "helper/template";
@include template.template;

.vp-reveal {
  position: relative;
  width: 100%;
  height: 400px;
  min-height: 400px;

  @media print {
    page-break-inside: avoid;
  }

  @media (max-width: hope-config.$mobile) {
    width: calc(100% + 3rem);
    min-height: 300px;
    margin: 0 -1.5rem;
  }
}

.reveal-loading {
  position: absolute;
  inset: 0 0 auto;
  color: var(--vp-tc);
}
