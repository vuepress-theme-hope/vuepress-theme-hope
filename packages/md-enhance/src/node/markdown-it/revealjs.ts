import { uml } from "@mdit/plugin-uml";
import { encodeData } from "@vuepress/helper";
import type { PluginSimple } from "markdown-it";

export const revealJs: PluginSimple = (md) => {
  md.use(uml, {
    name: "revealjs",
    open: "slidestart",
    close: "slideend",
    render: (tokens, index): string => {
      const token = tokens[index];
      const key = `revealjs-${index}`;
      const { content, info } = token;

      return `<RevealJs id="${key}" code="${encodeData(content)}" theme="${
        info.trim() || "auto"
      }"></RevealJs>`;
    },
  });
};
