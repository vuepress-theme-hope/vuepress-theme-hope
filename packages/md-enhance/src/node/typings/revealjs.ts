export type RevealJsPlugin = "highlight" | "math" | "search" | "notes" | "zoom";

export type RevealJsTheme =
  | "auto"
  | "beige"
  | "black"
  | "blood"
  | "league"
  | "moon"
  | "night"
  | "serif"
  | "simple"
  | "sky"
  | "solarized"
  | "white";

/**
 * Reveal.js options
 */
export interface RevealJsOptions {
  /**
   * Reveal.js plugins
   *
   * 幻灯片插件
   *
   * @default []
   */
  plugins?: RevealJsPlugin[];

  /**
   * Reveal.js themes
   *
   * 幻灯片主题
   *
   * @default ["auto"]
   */
  themes?: RevealJsTheme[];
}
