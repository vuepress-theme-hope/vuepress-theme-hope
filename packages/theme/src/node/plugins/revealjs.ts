import { isPlainObject } from "@vuepress/helper";
import type { RevealJsPluginOptions } from "@vuepress/plugin-revealjs";
import type { Plugin } from "vuepress/core";
import { colors } from "vuepress/utils";

import { logger } from "../utils.js";

let revealJsPlugin: ((options: RevealJsPluginOptions) => Plugin) | null = null;

try {
  ({ revealJsPlugin } = await import("@vuepress/plugin-revealjs"));
} catch {
  // Do nothing
}

/**
 * @private
 *
 * Resolve options for @vuepress/plugin-revealjs
 */
export const getRevealJsPlugin = (
  options?: Omit<RevealJsPluginOptions, "layout"> | boolean,
): Plugin | null => {
  if (!options) return null;

  if (!revealJsPlugin) {
    logger.error(
      `${colors.cyan("@vuepress/plugin-revealjs")} is not installed!`,
    );

    return null;
  }

  return revealJsPlugin(
    isPlainObject(options) ? { ...options, layout: "SlidePage" } : {},
  );
};
