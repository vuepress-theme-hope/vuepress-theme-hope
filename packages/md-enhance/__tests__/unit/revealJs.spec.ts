import MarkdownIt from "markdown-it";
import { describe, expect, it } from "vitest";

import { revealJs } from "../../src/node/markdown-it/revealjs.js";

const demo = `
## Slide 1

A paragraph with some text and a [link](https://mister-hope.com)

---

## Slide 2

- Item 1
- Item 2

---

## Slide 3.1

\`\`\`js
const a = 1;
\`\`\`

--

## Slide 3.2

$$
J(\\theta_0,\\theta_1) = \\sum_{i=0}
$$
`;

describe("revealJs", () => {
  const markdownIt = MarkdownIt({ linkify: true }).use(revealJs);

  it("Should render", () => {
    const renderResult = markdownIt.render(`
@slidestart
${demo}
@slideend
`);

    expect(renderResult).toMatch(
      /<RevealJs id="revealjs-.*?" code=".*?" theme=".*?"><\/RevealJs>/,
    );
    expect(renderResult).toMatchSnapshot();
  });

  it("Should not render", () => {
    expect(
      markdownIt.render(`
${demo}
`),
    ).toMatchSnapshot();

    expect(
      markdownIt.render(`
@slidestar
${demo}
@slideend
`),
    ).toMatchSnapshot();
  });
});
