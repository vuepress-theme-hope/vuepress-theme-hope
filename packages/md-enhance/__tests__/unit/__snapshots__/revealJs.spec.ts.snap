// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`revealJs > Should not render 1`] = `
"<h2>Slide 1</h2>
<p>A paragraph with some text and a <a href="https://mister-hope.com">link</a></p>
<hr>
<h2>Slide 2</h2>
<ul>
<li>Item 1</li>
<li>Item 2</li>
</ul>
<hr>
<h2>Slide 3.1</h2>
<pre><code class="language-js">const a = 1;
</code></pre>
<p>--</p>
<h2>Slide 3.2</h2>
<p>$$
J(\\theta_0,\\theta_1) = \\sum_{i=0}
$$</p>
"
`;

exports[`revealJs > Should not render 2`] = `
"<p>@slidestar</p>
<h2>Slide 1</h2>
<p>A paragraph with some text and a <a href="https://mister-hope.com">link</a></p>
<hr>
<h2>Slide 2</h2>
<ul>
<li>Item 1</li>
<li>Item 2</li>
</ul>
<hr>
<h2>Slide 3.1</h2>
<pre><code class="language-js">const a = 1;
</code></pre>
<p>--</p>
<h2>Slide 3.2</h2>
<p>$$
J(\\theta_0,\\theta_1) = \\sum_{i=0}
$$</p>
<p>@slideend</p>
"
`;

exports[`revealJs > Should render 1`] = `"<RevealJs id="revealjs-0" code="eJxVjb0KwjAUhff7FBfaoQXTH92UDo66OlppQxtMtGlCc0VBfHcTKJRuh4/vnANRhJdB9QJLgCNaPvH7xK3EtyKJzmiBJD6EfOyR43VQ4/OWSCLr9nmulSMxMWmsyDqjUwDGGCyTWw/wREL78TkEtHJ2mT9u2/bhoDOj809YYXkIKKgr05fjGM5JTVIQb4rNHMrUd2r30s1XVcUvSH9FOzuV" theme="auto"></RevealJs>"`;
